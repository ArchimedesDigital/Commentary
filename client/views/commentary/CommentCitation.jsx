import RaisedButton from 'material-ui/RaisedButton';
import Popover from 'material-ui/Popover';
import Menu from 'material-ui/Menu';
import MenuItem from 'material-ui/MenuItem';

CommentCitation = React.createClass({

	propTypes: {
		componentClass: React.PropTypes.string,
		title: React.PropTypes.string,
		comment: React.PropTypes.object,
	},

	getInitialState() {
		return {
			openMenu: false,
			anchorEl: null,
		};
	},

	handleTouchTap(event) {
		// This prevents ghost click.
		event.preventDefault();

		this.setState({
			openMenu: true,
			anchorEl: event.currentTarget,
		});
	},

	handleRequestClose() {
		this.setState({
			openMenu: false,
		});
	},

	render() {
		const { comment } = this.props;
		const title = this.props.title;
		const componentClass = this.props.componentClass;

		return (
			<div className={componentClass}>
				<RaisedButton
					label={title}
					labelPosition="after"
					onTouchTap={this.handleTouchTap}
				/>
				<Popover
					open={this.state.openMenu}
					anchorEl={this.state.anchorEl}
					anchorOrigin={{ horizontal: 'left', vertical: 'top' }}
					targetOrigin={{ horizontal: 'left', vertical: 'bottom' }}
					onRequestClose={this.handleRequestClose}
				>
					<Menu>
						{comment.revisions.map((revision, i) => {
							let format = 'D MMMM YYYY';
							let updated = revision.updated;

							if (revision.originalDate) {
								updated = revision.originalDate;
							}

							return (
								<MenuItem
									key={i}
									href={`/commentary/?_id=${comment._id}&revision=${i}`}
									primaryText={`Revision ${moment(updated).format(format)}`}
								/>
							);
						})}
						<MenuItem href={`/commentary/?_id=${comment._id}`} primaryText="Comment link" />
					</Menu>
				</Popover>
			</div>
		);
	},
});
