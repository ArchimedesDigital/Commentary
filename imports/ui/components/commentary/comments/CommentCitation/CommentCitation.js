import RaisedButton from 'material-ui/RaisedButton';
import Popover from 'material-ui/Popover';
import Menu from 'material-ui/Menu';
import MenuItem from 'material-ui/MenuItem';

const getDateRevision = (revision) => {
	if (revision.originalDate) return revision.originalDate;
	else if (revision.updated) return revision.updated;
	return revision.created;
};

class CommentCitation extends React.Component {

	constructor(props) {
		super(props);

		this.state = {
			openMenu: false,
			anchorEl: null,
		};
	}

	handleTouchTap(event) {
		// This prevents ghost click.
		event.preventDefault();

		this.setState({
			openMenu: true,
			anchorEl: event.currentTarget,
		});
	}

	handleRequestClose() {
		this.setState({
			openMenu: false,
		});
	}

	render() {

		const { commentId, revisions } = this.props;
		const { openMenu, anchorEl } = this.state;

		const styles = {
			menuItem: {
				fontFamily: 'ProximaNW01-AltLightReg',
			}
		};

		return (
			<div className="comment-citation">
				<RaisedButton
					label="Cite this comment"
					labelPosition="after"
					onClick={this.handleTouchTap.bind(this)}
				/>
				<Popover
					open={openMenu}
					anchorEl={anchorEl}
					anchorOrigin={{ horizontal: 'left', vertical: 'top' }}
					targetOrigin={{ horizontal: 'left', vertical: 'bottom' }}
					onRequestClose={this.handleRequestClose.bind(this)}
				>
					<Menu>
						{revisions.map((revision, i) => {

							const updated = getDateRevision(revision);

							return (
								<MenuItem
									key={revision._id}
									href={`/commentary/?_id=${commentId}&revision=${i}`}
									primaryText={`Revision ${moment(updated).format('D MMMM YYYY')}`}
									style={styles.menuItem}
								/>
							);
						})}
						<MenuItem
							href={`/commentary/?_id=${commentId}`}
							primaryText="Comment link"
							style={styles.menuItem}
						/>
					</Menu>
				</Popover>
			</div>
		);
	}
}
CommentCitation.propTypes = {
	commentId: React.PropTypes.string.isRequired,
	revisions: React.PropTypes.arrayOf(React.PropTypes.shape({
		_id: React.PropTypes.string.isRequired,
		created: React.PropTypes.instanceOf(Date),
		updated: React.PropTypes.instanceOf(Date),
		originalDate: React.PropTypes.instanceOf(Date),
	})).isRequired,
};

export default CommentCitation;
