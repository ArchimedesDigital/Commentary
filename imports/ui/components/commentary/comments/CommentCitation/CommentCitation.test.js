import React from 'react';
import renderer from 'react-test-renderer';

// component:
import CommentCitation from './CommentCitation';

describe('CommentCitation', () => {
	it('renders correctly', () => {

		const tree = renderer
			.create(<CommentCitation commentId="" revisions={[]} />)
			.toJSON();
		expect(tree).toMatchSnapshot();
	});
});

// TODO Fix error:
// Warning: Failed context type: The context `muiTheme` is marked as required in `RaisedButton`, but its value is `undefined`.
// 	in RaisedButton (created by CommentCitation)
// in div (created by CommentCitation)
// in CommentCitation

