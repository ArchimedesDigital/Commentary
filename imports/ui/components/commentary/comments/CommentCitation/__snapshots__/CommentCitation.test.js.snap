// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CommentCitation renders correctly 1`] = `null`;
